include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3681 @@
+    // singleton object, there is no need to keep summary object. For
+    // non-singleton, summary object in current state stores properties
+    object_value_t &odi_val = res_prod.second();
+    if (!crab_domain_params_man::get().singletons_in_base()) {
+      // Not necessary to reset cache because there is no ref_store/ref_load
+      // occur. However, we still need to commit cache.
+      reduce_regs_flds_between_cache_and_base(m_base_dom, m_eq_regs_dom,
+                                              odi_val, flds_vec);
+      // copy singleton object into summary
+      odi_val.first() = odi_val.second().first();
+      odi_val.second().second().set_to_top();
+    } else {
+      // TODO: performance may matter if the base dom contains huge dimensions
+      // instead of copy the base domain and perform project, it is better to
+      // have a special project onto operation returns a new domain after
+      // projection.
+      base_abstract_domain_t singleton_base(m_base_dom);
+      m_ghost_var_num_man.project(flds_vec, m_base_dom);
+
+      // copy singleton object into summary
+      odi_val.first() = singleton_base;
+      for (auto &fld : flds_vec) {
+        m_ghost_var_num_man.forget(fld, m_base_dom);
+      }
+    }
+    // update odi map
+    m_odi_map.set(id, res_prod);
+
+    CRAB_LOG("object-move-odi", crab::outs() << "move object: " << id << "\n";
+             m_odi_map.odi_write(crab::outs(), res_prod);
+             crab::outs() << "\n";);
+  }
+
+  /***************** Cache operations *****************/
+  // commit cache contents into summary
+  void commit_cache(field_abstract_domain_t &summary,
+                    field_abstract_domain_t &cache) const {
+    // join summary with cache
+    summary |= cache;
+  }
+
+  // update cache contents from summary to cache
+  void update_cache(field_abstract_domain_t &summary,
+                    field_abstract_domain_t &cache) const {
+    // copy summary to cache
+    cache = summary;
+  }
+
+  /// @brief commit cache contents into summary if it is dirty and reset
+  /// @param base_dom the base domain on current state
+  /// @param eq_regs_dom the equality domain for registers
+  /// @param prod object value stores a product of <SUM dom, cache dom, EQ dom>
+  /// @param obj_info_ref the object info
+  /// @param id an object id
+  void
+  commit_cache_if_dirty(base_abstract_domain_t &base_dom,
+                        const eq_register_abstract_domain_t &eq_regs_dom,
+                        object_value_t &prod,
+                        const object_domain_impl::object_info &obj_info_ref,
+                        const obj_id_t &id) const {
+    base_dom_variable_vector_t obj_flds;
+    get_obj_flds_with_ghost(id, obj_flds);
+    // (1) perform reduction by transfering regs-flds' constriants between cache
+    //  domain and base domain
+    reduce_regs_flds_between_cache_and_base(base_dom, eq_regs_dom, prod,
```

Comment:
is this required for soundness?

Reply:
No, I added this because we no longer keep equality between fields and registers once the cache is flushed. If we commented this out. We just lost numerical properties for equality (i.e. imprecise but still sound).